lib/Effect/Fs.hs

```diff
@@ -0,0 +1,179 @@
+  deriving stock (Show)
+newtype Directory a = Directory a
+  deriving stock (Show)
+
+data FsError =
+  FENodeNotFound
+  { fePath :: Path
+  }
+  | FENodeExists
+  { feWanted :: Node' Path
+  , feFound :: Node' Path
+  }
+  | FEMissingParent
+  { feNode :: Node' Path
+  }
+  | FEInvalidPath
+  { fePode :: Path
+  }
+  deriving stock (Show)
+
+newtype NodeRec = NodeRec (Node Path [NodeRec])
+
+pathToString :: Path -> Either FsError String
+pathToString path =
+  case decodeUtf8' path of
+    Left a -> Left $ FEInvalidPath path
+    Right b -> Right $ T.unpack b
+stringToPath :: String -> Path
+stringToPath = encodeUtf8 . T.pack
+
+extractNodePath
+  :: Node' Path
+  -> ByteString
+extractNodePath =
+  \case Left (File path) -> path
+        Right (Directory path) -> path
+
+eitherError
+  :: Member (Error e) r
+  => (a -> e)
+  -> Either a b
+  -> Sem r b
+eitherError f = either (throw . f) pure
+
+data FsEffect m a where
+  NodeExists :: Path -> FsEffect m (Maybe (Node' ()))
+  GetNode :: Path -> FsEffect m (Maybe (Node' Path))
+  ListDirectory :: Directory Path -> FsEffect m [Node' ByteString]
+  ListDirectoryRec :: Directory Path -> FsEffect m [NodeRec]
+--  ReadNode :: Node' Path -> FsEffect m (Node ByteString [ByteString])
+--  CreateNode :: Node' Path -> FsEffect m (Node' ())
+--  GetHandle :: File Path -> FsEffect m (File Handle)
+makeSem ''FsEffect
+
+runFsInIO
+  :: Member (Error FsError) r
+  => Member (Embed IO) r
+  => Sem (FsEffect ': r) a
+  -> Sem r a
+runFsInIO = interpret
+  \case
+    NodeExists path -> _nodeExists path
+    GetNode path -> _getNode path
+    ListDirectory dirPath -> _listDirectory dirPath
+    ListDirectoryRec dirPath -> _listDirectoryRec dirPath
+    -- ReadNode nodePath -> undefined
+    -- CreateNode nodePath -> undefined
+
+_nodeExists
+  :: Member (Error FsError) r
+  => Member (Embed IO) r
+  => Path
+  -> Sem r (Maybe (Node' ()))
+_nodeExists path = do
+  stringPath <-
+    eitherError
+    (const $ FEInvalidPath path)
+    (decodeUtf8' path <&> T.unpack)
+  file <- embed $ doesFileExist stringPath
+  dir <- embed $ doesDirectoryExist stringPath
+
+  case (file, dir) of
+    (True, False) -> pure . Just . Left $ File ()
+    (False, True) -> pure . Just . Right $ Directory ()
+    (_, _) -> pure Nothing
+
+_getNode
+  :: Member (Error FsError) r
+  => Member (Embed IO) r
+  => Path
+  -> Sem r (Maybe (Node' Path))
+_getNode path =
+  _nodeExists path
+  <&> \case
+        Just (Left _) -> Just (Left (File path))
+        Just (Right _) -> Just (Right (Directory path))
+        Nothing -> Nothing
+
+_listDirectory
+  :: Member (Error FsError) r
+  => Member (Embed IO) r
+  => Directory Path
+  -> Sem r [Node' ByteString]
+_listDirectory (Directory path) = do
+  stringPath <- eitherError id (pathToString path)
+  nodes <- embed $ D.listDirectory stringPath
+  mapM (_getNodeThrow . stringToPath) nodes
+    where
+      _getNodeThrow
+        :: Member (Error FsError) r
+        => Member (Embed IO) r
```

Comment:
Same here (`ScopedTypeVariables`).

Reply:
It does give me a warning though

Reply:
What warning? :thinking:

Reply:
I have no idea, I'll look for it, disregard this for now

Reply:
resolved everywhere i think, ill check

Reply:
I meant that you can write `forall r.` in `_listDirectory` type and delete these constraints in `_getNodeThrow`. Not to delete `_getNodeThrow` type.